File: JavaScript/930-Binary Subarrays With Sum.js
/**
 * @param {number[]} A
 * @param {number} S
 * @return {number}
 */
var numSubarraysWithSum = function(A, S) {
    let preSum = new Map();
    preSum.set(0, 1);
    let total = 0, res = 0;
    
    for (let num of A){
        total += num;
        if (preSum.has(total-S)){
            res += preSum.get(total-S);
        }
        preSum.set(total, (preSum.get(total) || 0) + 1 );
    }
    return res;
};

File: C++/930-Binary Subarrays With Sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        int cnt = 0, res = 0;
        vector<int> zeroArr;
        
        for (int num:nums) {
            if (num == 0) cnt++;
            else {
                zeroArr.push_back(cnt);
                cnt = 0;
            }
        }
        zeroArr.push_back(cnt);

        if (goal > 0) {
            for (int i=0; (i+goal)<zeroArr.size(); i++) {
                res += (zeroArr[i]+1)*(zeroArr[i+goal]+1);
            }
        } else {
            for (int n:zeroArr) {
                res += n*(n+1)/2;
            }
        }
        
        return res;
    }
};